reroute/src/Web/Routing/SafeRouting.hs

```diff
@@ -125,67 +125,49 @@ insertPathMap (RouteHandle path action) = insertPathMap' path (HV.uncurry action
 
 insertSubComponent' :: Path ts -> (HVect ts -> [T.Text] -> x) -> PathMap x -> PathMap x
 insertSubComponent' path subComponent =
-  let updateSubComponents y (PathMap c h s p) = PathMap (y : c) h s p
+  let updateSubComponents y (PathMap c h s p w) = PathMap (y : c) h s p w
   in updatePathMap updateSubComponents path subComponent
 
 insertSubComponent :: Functor m => RouteHandle m ([T.Text] -> a) -> PathMap (m a) -> PathMap (m a)
 insertSubComponent (RouteHandle path comp) =
   insertSubComponent' path (fmap (\m ps -> fmap ($ ps) m) (HV.uncurry comp))
 
 match :: PathMap x -> [T.Text] -> [x]
-match (PathMap c h s p) pieces =
+match (PathMap c h s p w) pieces =
   map ($ pieces) c ++
   case pieces of
-    [] -> h
+    [] -> h ++ fmap ($ "") w
     (pp:pps) ->
       let staticMatches = maybeToList (HM.lookup pp s) >>= flip match pps
           varMatches = PM.lookupConcat (fromPathPiece pp)
                          (\piece pathMap' -> fmap ($ piece) (match pathMap' pps)) p
-      in staticMatches ++ varMatches
-
--- | A route parameter
-var :: (Typeable a, PathPiece a) => Path (a ': '[])
-var = VarCons Empty
-
-type Var a = Path (a ': '[])
-
--- | A static route piece
-static :: String -> Path '[]
-static s =
-  let pieces = filter (not . T.null) $ T.splitOn "/" $ T.pack s
-  in foldr StaticCons Empty pieces
-
-instance (a ~ '[]) => IsString (Path a) where
-    fromString = static
-
--- | The root of a path piece. Use to define a handler for "/"
-root :: Path '[]
-root = Empty
-
-(</>) :: Path as -> Path bs -> Path (Append as bs)
-(</>) Empty xs = xs
-(</>) (StaticCons pathPiece xs) ys = (StaticCons pathPiece (xs </> ys))
-(</>) (VarCons xs) ys = (VarCons (xs </> ys))
+          routeRest = combineRoutePieces pieces
+          wildcardMatches = fmap ($ routeRest) w
+      in staticMatches ++ varMatches ++ wildcardMatches
 
 renderRoute :: Path as -> HVect as -> T.Text
-renderRoute p h =
-    T.intercalate "/" $ renderRoute' p h
+renderRoute p = combineRoutePieces . renderRoute' p
 
 renderRoute' :: Path as -> HVect as -> [T.Text]
 renderRoute' Empty _ = []
 renderRoute' (StaticCons pathPiece pathXs) paramXs =
     ( pathPiece : renderRoute' pathXs paramXs )
 renderRoute' (VarCons pathXs) (val :&: paramXs) =
     ( toPathPiece val : renderRoute' pathXs paramXs)
+renderRoute' (Wildcard pathXs) (_ :&: paramsXs) =
```

Comment:
A user should only be able to render `Path`s without wildcards, as this can be used to generate (type safe) hyperlinks and this would break it.